### PR TITLE
Add I3D Chisel Generator

### PR DESCRIPTION
**Pull request by harrisonliew**
Open-sourcing the I3D Chisel generator.
Though the link has been lost, it was forked from the [Chisel project template](https://github.com/chipsalliance/chisel-template).

### CODE REVIEW

File: i3d/src/main/scala/Collateral.scala [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+      ("vlsi.inputs.supplies.ground" -> Seq(("name" -> "VSS") ~ ("pins" -> Seq("VSS"))))
+
+    pretty(render(bumps merge pins merge places merge sdc merge power))
+  }
+
+  def toSDC(shiftCase: Boolean = false): String = {
+    // Extract bump objects
+    val bumps = iocells.map(_.forBump)
+
+    // TODO: power is bump pitch related.
```
Comment (swallenintel) [inline i3d/src/main/scala/Collateral.scala:195]: Is this TODO important for this PR? Should it be tracked as an issue?

File: i3d/src/main/scala/Collateral.scala [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+    val globalSDC = f"""# Global constraints
+      |# All timing units assumed in ns
+      |# Loads/transitions (config pins, internal)
+      |set_load 0.005 [all_outputs]
+      |set_max_transition 0.05 [current_design]
+      |# Power. Not supported by all tools.
+      |set_max_dynamic_power ${0.05 * p.numMods * p.sigsPerMod / p.tPeriod}%.1f mW [current_design]
+      |# False paths from config pins
+      |set_false_path -from *ioCtrl*
```
Comment (swallenintel) [inline i3d/src/main/scala/Collateral.scala:204]: I don't like `ioCtrl` as a false path targeting string. That could easily get used for unrelated code.

This string should be very verbose, see the example I showed you the other day!

File: i3d/src/main/scala/Collateral.scala [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+  def toSDC(shiftCase: Boolean = false): String = {
+    // Extract bump objects
+    val bumps = iocells.map(_.forBump)
+
+    // TODO: power is bump pitch related.
+    val globalSDC = f"""# Global constraints
+      |# All timing units assumed in ns
+      |# Loads/transitions (config pins, internal)
+      |set_load 0.005 [all_outputs]
+      |set_max_transition 0.05 [current_design]
```
Comment (swallenintel) [inline i3d/src/main/scala/Collateral.scala:200]: This value can be technology specific. Is this SDC supposed to be user editable after generation?

If not, this needs to be configured somehow.

File: i3d/src/main/scala/Collateral.scala (hunk 1 above)
Comment (swallenintel) [reply i3d/src/main/scala/Collateral.scala:195]: Other TODOs in this doc too. Same goes for all of them.

File: i3d/src/main/scala/Collateral.scala [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+          s"${tx.bumpName} io_${tx.bumpName} out_${tx.bumpName}"
+        } ++
+        rxClks.map{ rx =>
+          s"${rx.bumpName} io_${rx.bumpName} out_${rx.bumpName}"
+        }
+    }
+    val globalClkSDC = Seq(
+      "# Config clock",
+      "create_clock clock -name clock -period 1.0",
+      "set_clock_uncertainty 0.02 [get_clocks clock]",
```
Comment (swallenintel) [inline i3d/src/main/scala/Collateral.scala:282]: Similar to max tran, this value is tricky. Is this user configurable?

File: i3d/src/main/scala/Utils.scala [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+    for (r <- 0 until rows; c <- 0 until cols)
+      // Update coordinates using calculated pitch and bumpOffset.
+      bumpMap(r)(c).location = I3DCoordinates[Double](
+        x = (c + 0.5) * gp.pitchH + (if (pinSide == "W") ip.bumpOffset else 0.0),
+        y = (r + 0.5) * gp.pitchV + (if (pinSide == "S") ip.bumpOffset else 0.0)
+      )
+
+    // Calculate pin locations
+    // Check for available routing resources
+    // Assume 20% power assumption + 50% shielding/space penalty
```
Comment (swallenintel) [inline i3d/src/main/scala/Utils.scala:505]: Is this assumption valid for all targeted technologies?

**Review by swallenintel — APPROVED** [review]

**Review by xinyang2k — APPROVED** [review]
